program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
+                    0
+                } else {
+                    self.valid_entry_counter += 1;
+                    1
+                };
+
+                //update max_update_time
+                self.max_update_time_gap = update_time;
+
+
+                self.add_price_to_entry(
+                    inflated_prev_price,
+                    inflated_prev_conf,
+                    inflated_current_price,
+                    inflated_current_conf,
+                    update_time,
+                    current_entry,
+                )?;
+            }
+            _ => {
+                //invalidate all the entries in your way
+                //this is ok because THRESHOLD < Granuality
+                self.invaldate_following_entries(
+                    cmp::min(num_skipped_entries, NUM_ENTRIES),
+                    prev_entry,
+                );
+                self.max_update_time_gap = update_time;
+            }
+        }
+        Ok(())
+    }
+}
+
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an Tick Tracker that has NUM_ENTRIES entries
+/// each tracking the last tick before every GRANUALITY seconds.
+pub struct TickTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    prices:         [SignedTrackerRunningSum; NUM_ENTRIES], /* price running time
+                                                             * weighted sums */
+    confidences:    [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                               * time
+                                                               * weighted sums */
+    entry_validity: [u8; NUM_ENTRIES], //entry validity
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    ///invalidates current_entry, and increments self.current_entry num_entries times
+    fn invaldate_following_entries(&mut self, num_entries: usize, current_entry: usize) {
+        let mut entry = current_entry;
+        for _ in 0..num_entries {
+            self.entry_validity[entry] = 0;
+            entry = Self::get_next_entry(entry);
+        }
+    }
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    Tracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+    for TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    fn initialize(&mut self) -> Result<(), OracleError> {
+        Ok(())
+    }
 
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        let prev_entry = Self::time_to_entry(prev_time)?;
+        let current_entry = Self::time_to_entry(current_time)?;
+        let num_skipped_entries = current_entry - prev_entry;
+        match num_skipped_entries {
```

Comment:
i find it somewhat confusing that we don't interpolate the prices between the two times in this case, but we do in the SMA case above. Is there a good reason for that? If not, I would suggest picking one or the other in both places.

Reply:
interesting, I think I asked Jeff about both and he seemed to agree that they are "robust", I can check with him if we want to interpolate, but that would change the validity condition.

This, Tick, and SMA, is what I recall agreeing in with atleast two other people in team before doing this. so it sounds like we need to have a discussion about what exactly we want to provide